test/test_connectionpool.py

```diff
@@ -226,6 +226,45 @@ def test_contextmanager(self):
 
         self.assertRaises(Empty, old_pool_queue.get, block=False)
 
+    def test_connection_recycling(self):
+        '''
+        Test that connections are recycled to the pool on 
+        connection/readtimeout errors where no http response 
+        is received.
+        '''
+        poolsize = 2
+        with HTTPConnectionPool('google.com', maxsize=poolsize, 
+                                  retries=1, block=True) as pool:
+
+            self.assertEqual(pool.pool.qsize(), poolsize)
+
+            # set a silly read timeout to force timeout exception. 
+            # We won't get a response for this and so the conn won't 
+            # be implicitly returned to the pool.
+            pool.timeout._read = 0
+            self.assertRaises(MaxRetryError, pool.urlopen, 'GET', 
```

Comment:
Whatever works for you. Feel free to piggyback on any existing tests too, if we were already exercising the condition but just weren't asserting the pool size.

But anything that actually opens a socket should go in with_dummyserver subdir.

Note that all the tests in this suite might setup pools but none of them actually make requests.